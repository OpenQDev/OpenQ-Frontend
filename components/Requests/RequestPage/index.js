import React, { useContext, useState } from 'react';
import RequestIndividual from '../RequestIndividual/index.js';
import { getPlural } from '../../../services/utils/lib';
import StoreContext from '../../../store/Store/StoreContext';
import AuthContext from '../../../store/AuthStore/AuthContext';
import PaginatedList from '../../Utils/PaginatedList/index.js';
import { fetchRequestsWithServiceArg } from '../../../services/utils/lib';

// Custom
import { useRouter } from 'next/router';
import useWeb3 from '../../../hooks/useWeb3.js';

const RequestPage = () => {
  const router = useRouter();
  const [authState] = useContext(AuthContext);
  const [appState] = useContext(StoreContext);
  const { accountData } = appState;
  const { gnosisSafe } = useWeb3();

  const loggedId = accountData?.id;
  const userId = router.query.userId;

  const isOwner = loggedId == userId;
  const { githubId, email } = authState;
  const getItems = async (oldCursor, batch, ordering = 'asc', filters = {}) => {
    try {
      return await fetchRequestsWithServiceArg(appState, identity, oldCursor, batch, ordering, filters);
    } catch (e) {
      appState.logger.error(e);
      return { nodes: [], cursor: null, complete: true };
    }
  };
  const identity = { userId, githubId, email };
  const filterFunction = () => {
    return true;
  };
  const paginationObj = {
    items: [],
    ordering: { direction: 'asc', field: 'name' },
    fetchFilters: {},
    filters: {},
    filterFunction,
    cursor: null,
    complete: false,
    batch: 10,
    getItems,
  };
  const paginationState = useState(paginationObj);
  const requestsLength = paginationState[0]?.items.length;

  return (
    <>
      <div className='my-6'>
        <h2 className='text-2xl font-semibold pb-4 border-b border-web-gray my-4'>Manage your bounties</h2>
        <div className='border-web-gray border flex justify-center content-center h-24 rounded-sm items-center'>
          You have received {requestsLength} request{getPlural(requestsLength)}.
        </div>
      </div>
      <div className='my-6'>
        <h2 className='text-2xl font-semibold pb-4  my-4'>Requests</h2>
        <div className='bg-info border-info-strong border-2 p-3 rounded-sm my-4'>
          If you accept the request, the amount deposited in the contract will be unlocked for the builder. If you are
          not satisfied with the work, please communicate with the builder, and accept the request once you are
          satisfied with their submission.
        </div>

        {gnosisSafe && (
          <div className='bg-info border-info-strong border-2 p-3 rounded-sm my-4'>
            Hey! Looks like you are using gnosis safe via wallet connect. Because gnosis safes often require multiple
            signatures, this modal will will be stuck in a pending state. Once you're multi-sig has approved the
            transaction, please reload the app, and you'll see the results of your transaction.
          </div>
        )}
        <ul className='flex flex-col gap-4'>
          {isOwner && <PaginatedList paginationState={paginationState} PaginationCard={RequestIndividual} />}
        </ul>
      </div>
    </>
  );
};
export default RequestPage;
