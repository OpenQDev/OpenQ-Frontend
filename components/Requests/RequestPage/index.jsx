import React, { useContext, useState } from 'react';
import RequestIndividual from '../RequestIndividual';
import { getPlural } from '../../../services/utils/lib';
import StoreContext from '../../../store/Store/StoreContext';
import AuthContext from '../../../store/AuthStore/AuthContext';
import PaginatedList from '../../Utils/PaginatedList';
import { fetchRequestsWithServiceArg } from '../../../services/utils/lib';
import SearchBar from '../../Search/SearchBar';
import { useRouter } from 'next/router';
import useWeb3 from '../../../hooks/useWeb3';

const RequestPage = ({ states }) => {
  const router = useRouter();
  const [authState] = useContext(AuthContext);
  const [appState] = useContext(StoreContext);
  const { accountData } = appState;
  const { gnosisSafe } = useWeb3();
  const [searchText, setSearchText] = useState('');

  const loggedId = accountData?.id;
  const userId = router.query.userId;

  const isOwner = loggedId == userId;
  const { githubId, email } = authState;
  const getItems = async (oldCursor, batch, ordering = 'asc', fetchFilters = {}) => {
    try {
      return await fetchRequestsWithServiceArg(appState, identity, oldCursor, batch, ordering, fetchFilters);
    } catch (e) {
      appState.logger.error(e);
      return { nodes: [], cursor: null, complete: true };
    }
  };
  const identity = { userId, githubId, email };
  const filterFunction = (item, filters) => {
    if (!filters?.searchText) return true;
    else return item.request.requestingUser?.githubUser.login.toLowerCase().includes(filters.searchText.toLowerCase());
  };
  const paginationObj = {
    items: [],
    ordering: { direction: 'asc', field: 'name' },
    fetchFilters: { states },
    filters: { searchText: '' },
    filterFunction,
    cursor: null,
    complete: false,
    batch: 20,
    getItems,
  };
  const paginationState = useState(paginationObj);
  const [paginationStateObj, setPaginationStateObj] = paginationState;
  const requestsLength = paginationState[0]?.items.length;
  const handleSearch = (e) => {
    setSearchText(e.target.value);
    setPaginationStateObj({
      ...paginationStateObj,
      filters: { ...paginationStateObj.filters, searchText: e.target.value },
    });
  };
  return (
    <>
      <div className='my-6'>
        <h2 className='text-2xl font-semibold pb-4 border-b border-web-gray my-4'>Manage your bounties</h2>
        <div className='border-web-gray border flex justify-center content-center h-24 rounded-sm items-center'>
          You have received {requestsLength} request{getPlural(requestsLength)}.
        </div>
      </div>
      <div className='my-6'>
        <h2 className='text-2xl font-semibold pb-4  my-4'>Requests</h2>
        <div className='bg-info border-info-strong border-2 p-3 rounded-sm my-4'>
          If the submission requires changes, please "Decline" the form submission and write a message to the builder in
          the popup. Once the submission is completed correctly, please click "Accept". If you accept the request, the
          amount deposited in the contract will be unlocked for the builder.
        </div>
        {gnosisSafe && (
          <div className='bg-info border-info-strong border-2 p-3 rounded-sm my-4'>
            Hey! Looks like you are using Gnosis Safe via WalletConnect. Because Gnosis Safes often require multiple
            signatures, this modal will will be stuck in a pending state. Once you're multisig has approved the
            transaction, please reload the app, and you'll see the results of your transaction.
          </div>
        )}
        <SearchBar placeholder='github-user' styles='w-full' value={searchText} onKeyUp={handleSearch} />
        <ul className='flex flex-col gap-4'>
          {isOwner && <PaginatedList paginationState={paginationState} PaginationCard={RequestIndividual} />}
        </ul>
      </div>
    </>
  );
};
export default RequestPage;
