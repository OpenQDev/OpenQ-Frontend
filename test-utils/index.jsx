// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
