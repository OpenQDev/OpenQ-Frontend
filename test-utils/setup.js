import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
