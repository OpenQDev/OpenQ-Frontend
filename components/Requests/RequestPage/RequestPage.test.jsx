import { render } from '../../../test-utils';
import React from 'react';
import RequestPage from '.';
import nextRouter from 'next/router';

nextRouter.useRouter = vi.fn();
nextRouter.useRouter.mockImplementation(() => ({
  query: { type: null },

  prefetch: vi.fn(() => {
    return { catch: vi.fn };
  }),
}));

describe('RequestPage', () => {
  it('should match initial DOM Snapshot', () => {
    const { asFragment } = render(<RequestPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
