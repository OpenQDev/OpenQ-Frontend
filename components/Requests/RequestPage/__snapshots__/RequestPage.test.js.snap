// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestPage should match initial DOM Snapshot 1`] = `
<DocumentFragment>
  <div
    class="my-6"
  >
    <h2
      class="text-2xl font-semibold pb-4 border-b border-web-gray my-4"
    >
      Manage your bounties
    </h2>
    <div
      class="border-web-gray border flex justify-center content-center h-24 rounded-sm items-center"
    >
      You have received 0 requests.
    </div>
  </div>
  <div
    class="my-6"
  >
    <h2
      class="text-2xl font-semibold pb-4  my-4"
    >
      Requests
    </h2>
    <div
      class="bg-info border-info-strong border-2 p-3 rounded-sm my-4"
    >
      If the submission requires changes, please "Decline" the form submission and write a message to the builder in the popup. Once the submission is completed correctly, please click "Accept". If you accept the request, the amount deposited in the contract will be unlocked for the builder.
    </div>
    <input
      class="flex-1 lg:col-span-3 col-span-4 input-field-big w-full"
      placeholder="github-user"
      type="text"
      value=""
    />
    <ul
      class="flex flex-col gap-4"
    >
      <div />
    </ul>
  </div>
</DocumentFragment>
`;
